saleor/graphql/core/connection.py

```diff
@@ -1,11 +1,228 @@
-from typing import Union
+import typing
 
 import graphene
 import opentracing as ot
-from django.db.models import Manager, QuerySet
+from django.db.models import Manager, Model as DjangoModel, Q, QuerySet
 from graphene import Field, List, NonNull, ObjectType, String
 from graphene.relay.connection import Connection
 from graphene_django_optimizer.types import OptimizedDjangoObjectType
+from graphql_relay.connection.connectiontypes import Edge, PageInfo
+from graphql_relay.utils import base64, unbase64
+
+from ..core.enums import OrderDirection
+
+ConnectionArguments = typing.Dict[str, typing.Any]
+
+
+def to_global_cursor(values):
+    if not isinstance(values, list):
+        values = [str(values)]
+    else:
+        values = [str(value) for value in values]
+    return base64("::".join(values))
+
+
+def from_global_cursor(cursor) -> typing.List[str]:
+    values = unbase64(cursor)
+    values = values.split("::")
+    if isinstance(values, list):
+        values = [None if value == "None" else value for value in values]
+        return values
+    if values == "None":
+        values = None
+    return [values]
+
+
+def get_field_value(instance: DjangoModel, field_name: str):
+    """Get field value for given field in filter format 'field__foreign_key_field'."""
+    field_path = field_name.split("__")
+    attr = instance
+    for elem in field_path:
+        attr = getattr(attr, elem)
+
+    if callable(attr):
+        return "%s" % attr()
+    return attr
+
+
+def _prepare_filter_expression(
+    field_name: str,
+    index: int,
+    cursor: typing.List,
+    sorting_fields: typing.List,
+    sorting_direction: str,
+) -> typing.Tuple:
```

Comment:
`List` and `Tuple` are generics that should be further qualified with type parameters

Reply:
done